File: src/game/alien_compat.c
// SNES alien offset compatibility layer.
// See alien_compat.h for details.

#include "alien_compat.h"

static bool alien_base_byte_ptr(Alien *al, uint16 offset, uint8 **out) {
    switch (offset) {
    case 4:  *out = ((uint8 *)&al->shape) + 0; return true;
    case 5:  *out = ((uint8 *)&al->shape) + 1; return true;
    case 6:  *out = ((uint8 *)&al->ptr) + 0; return true;
    case 7:  *out = ((uint8 *)&al->ptr) + 1; return true;
    case 8:  *out = &al->flags; return true;
    case 9:  *out = &al->type; return true;
    case 10: *out = &al->count; return true;
    case 11: *out = &al->count1; return true;
    case 12: *out = ((uint8 *)&al->worldx) + 0; return true;
    case 13: *out = ((uint8 *)&al->worldx) + 1; return true;
    case 14: *out = ((uint8 *)&al->worldy) + 0; return true;
    case 15: *out = ((uint8 *)&al->worldy) + 1; return true;
    case 16: *out = ((uint8 *)&al->worldz) + 0; return true;
    case 17: *out = ((uint8 *)&al->worldz) + 1; return true;
    case 18: *out = &al->rotx; return true;
    case 19: *out = &al->roty; return true;
    case 20: *out = &al->rotz; return true;
    case 21: *out = &al->vel; return true;
    // al_stratptr (22..24 on SNES) is a 24-bit function pointer.
    // Flat-memory C function pointers are not byte-compatible; deny direct writes.
    case 25: *out = ((uint8 *)&al->immuneptr) + 0; return true;
    case 26: *out = ((uint8 *)&al->immuneptr) + 1; return true;
    case 27: *out = ((uint8 *)&al->collobjptr) + 0; return true;
    case 28: *out = ((uint8 *)&al->collobjptr) + 1; return true;
    case 29: *out = &al->sflags; return true;
    case 30: *out = &al->sflags2; return true;
    case 31: *out = &al->sflags3; return true;
    case 32: *out = &al->sflags4; return true;
    case 33: *out = &al->skidy; return true;
    case 34: *out = &al->sbyte1; return true;
    case 35: *out = &al->sbyte2; return true;
    case 36: *out = &al->sbyte3; return true;
    case 37: *out = &al->sbyte4; return true;
    case 38: *out = ((uint8 *)&al->sword1) + 0; return true;
    case 39: *out = ((uint8 *)&al->sword1) + 1; return true;
    case 40: *out = ((uint8 *)&al->sword2) + 0; return true;
    case 41: *out = ((uint8 *)&al->sword2) + 1; return true;
    case 42: *out = &al->HP; return true;
    case 43: *out = &al->AP; return true;
    default:
        return false;
    }
}

static bool alien_alx_byte_ptr(Alien *al, uint16 offset, uint8 **out) {
    switch (offset) {
    case 0:  *out = ((uint8 *)&al->sWPx1) + 0; return true;
    case 1:  *out = ((uint8 *)&al->sWPx1) + 1; return true;
    case 2:  *out = ((uint8 *)&al->sWPy1) + 0; return true;
    case 3:  *out = ((uint8 *)&al->sWPy1) + 1; return true;
    case 4:  *out = ((uint8 *)&al->sWPz1) + 0; return true;
    case 5:  *out = ((uint8 *)&al->sWPz1) + 1; return true;
    // alx strategy pointers (6..17 on SNES) are 24-bit function pointers.
    // Flat-memory C function pointers are not byte-compatible; deny direct access.
    case 18: *out = &al->stratstate; return true;
    case 19: *out = ((uint8 *)&al->fireobjptr) + 0; return true;
    case 20: *out = ((uint8 *)&al->fireobjptr) + 1; return true;
    case 21: *out = ((uint8 *)&al->depthoffset) + 0; return true;
    case 22: *out = ((uint8 *)&al->depthoffset) + 1; return true;
    case 23: *out = &al->relposx; return true;
    case 24: *out = &al->relposy; return true;
    case 25: *out = &al->relposz; return true;
    case 26: *out = ((uint8 *)&al->debrisshape) + 0; return true;
    case 27: *out = ((uint8 *)&al->debrisshape) + 1; return true;
    case 28: *out = &al->colframe; return true;
    case 29: *out = &al->animframe; return true;
    case 30: *out = &al->snd1; return true;
    case 31: *out = &al->snd2; return true;
    case 32: *out = ((uint8 *)&al->coltab) + 0; return true;
    case 33: *out = ((uint8 *)&al->coltab) + 1; return true;
    case 34: *out = &al->childx; return true;
    case 35: *out = &al->childy; return true;
    case 36: *out = &al->childz; return true;
    case 37: *out = &al->childrotx; return true;
    case 38: *out = &al->childroty; return true;
    case 39: *out = &al->childrotz; return true;
    case 40: *out = ((uint8 *)&al->childrotobj) + 0; return true;
    case 41: *out = ((uint8 *)&al->childrotobj) + 1; return true;
    case 42: *out = &al->tx; return true;
    case 43: *out = &al->ty; return true;
    case 44: *out = ((uint8 *)&al->memptr) + 0; return true;
    case 45: *out = ((uint8 *)&al->memptr) + 1; return true;
    case 46: *out = ((uint8 *)&al->stackptr) + 0; return true;
    case 47: *out = ((uint8 *)&al->stackptr) + 1; return true;
    case 48: *out = ((uint8 *)&al->stratmem) + 0; return true;
    case 49: *out = ((uint8 *)&al->stratmem) + 1; return true;
    case 50: *out = &al->pbyte1; return true;
    case 51: *out = &al->pbyte2; return true;
    case 52: *out = ((uint8 *)&al->pword1) + 0; return true;
    case 53: *out = ((uint8 *)&al->pword1) + 1; return true;
    default:
        return false;
    }
}

static bool alien_byte_ptr(Alien *al, uint16 offset, bool is_alx, uint8 **out) {
    return is_alx ? alien_alx_byte_ptr(al, offset, out)
                  : alien_base_byte_ptr(al, offset, out);
}

bool AlienCompat_Read8(const Alien *al, uint16 offset, bool is_alx, uint8 *out) {
    if (!al || !out) return false;
    uint8 *ptr = NULL;
    if (!alien_byte_ptr((Alien *)al, offset, is_alx, &ptr) || !ptr) return false;
    *out = *ptr;
    return true;
}

bool AlienCompat_Read16(const Alien *al, uint16 offset, bool is_alx, uint16 *out) {
    if (!al || !out) return false;
    uint8 lo = 0;
    uint8 hi = 0;
    if (!AlienCompat_Read8(al, offset, is_alx, &lo)) return false;
    if (!AlienCompat_Read8(al, (uint16)(offset + 1), is_alx, &hi)) return false;
    *out = (uint16)lo | ((uint16)hi << 8);
    return true;
}

bool AlienCompat_Write8(Alien *al, uint16 offset, bool is_alx, uint8 value) {
    if (!al) return false;
    uint8 *ptr = NULL;
    if (!alien_byte_ptr(al, offset, is_alx, &ptr) || !ptr) return false;
    *ptr = value;
    return true;
}

bool AlienCompat_Write16(Alien *al, uint16 offset, bool is_alx, uint16 value) {
    if (!al) return false;
    if (!AlienCompat_Write8(al, offset, is_alx, (uint8)(value & 0xFFu))) return false;
    if (!AlienCompat_Write8(al, (uint16)(offset + 1), is_alx, (uint8)(value >> 8))) return false;
    return true;
}

bool AlienCompat_Add8(Alien *al, uint16 offset, bool is_alx, int8 delta) {
    uint8 cur = 0;
    if (!AlienCompat_Read8(al, offset, is_alx, &cur)) return false;
    cur = (uint8)(cur + delta);
    return AlienCompat_Write8(al, offset, is_alx, cur);
}

bool AlienCompat_Add16(Alien *al, uint16 offset, bool is_alx, int16 delta) {
    uint16 cur = 0;
    if (!AlienCompat_Read16(al, offset, is_alx, &cur)) return false;
    cur = (uint16)((int16)cur + delta);
    return AlienCompat_Write16(al, offset, is_alx, cur);
}

File: rust/sf-path/src/alien_compat.rs
//! SNES alien offset compatibility layer.
//!
//! C origin: `src/game/alien_compat.c` / `alien_compat.h`. Maps 16-bit-era
//! `al_`/`alx_` byte offsets used by map/path bytecode onto the flat [`Alien`]
//! struct fields.
//!
//! When `is_alx` is false the offset is relative to `al_start` (includes the
//! 4-byte list header, so real fields start at offset 4). When `is_alx` is
//! true the offset is relative to `alx_start`.
//!
//! The C code hands out raw byte pointers into the struct
//! (`alien_base_byte_ptr` / `alien_alx_byte_ptr`); here each byte slot is a
//! get/set pair over the typed fields, preserving little-endian lo/hi
//! semantics for the 16-bit fields. Offsets covering 24-bit SNES function
//! pointers are denied exactly like the C layer (`al_stratptr` 22..24,
//! alx 6..17).

use crate::alien::Alien;

/// Read one byte of an `al_start`-relative slot (C `alien_base_byte_ptr` read).
fn base_read(al: &Alien, offset: u16) -> Option<u8> {
    Some(match offset {
        4 => al.shape.to_le_bytes()[0],
        5 => al.shape.to_le_bytes()[1],
        6 => al.ptr.to_le_bytes()[0],
        7 => al.ptr.to_le_bytes()[1],
        8 => al.flags,
        9 => al.type_,
        10 => al.count,
        11 => al.count1,
        12 => al.worldx.to_le_bytes()[0],
        13 => al.worldx.to_le_bytes()[1],
        14 => al.worldy.to_le_bytes()[0],
        15 => al.worldy.to_le_bytes()[1],
        16 => al.worldz.to_le_bytes()[0],
        17 => al.worldz.to_le_bytes()[1],
        18 => al.rotx,
        19 => al.roty,
        20 => al.rotz,
        21 => al.vel,
        // al_stratptr (22..24 on SNES) is a 24-bit function pointer; denied.
        25 => al.immuneptr.to_le_bytes()[0],
        26 => al.immuneptr.to_le_bytes()[1],
        27 => al.collobjptr.to_le_bytes()[0],
        28 => al.collobjptr.to_le_bytes()[1],
        29 => al.sflags,
        30 => al.sflags2,
        31 => al.sflags3,
        32 => al.sflags4,
        33 => al.skidy,
        34 => al.sbyte1,
        35 => al.sbyte2,
        36 => al.sbyte3,
        37 => al.sbyte4,
        38 => al.sword1.to_le_bytes()[0],
        39 => al.sword1.to_le_bytes()[1],
        40 => al.sword2.to_le_bytes()[0],
        41 => al.sword2.to_le_bytes()[1],
        42 => al.hp,
        43 => al.ap,
        _ => return None,
    })
}

/// Write one byte of an `al_start`-relative slot (C `alien_base_byte_ptr` write).
fn base_write(al: &mut Alien, offset: u16, value: u8) -> bool {
    fn set_lo(word: &mut u16, v: u8) {
        *word = (*word & 0xFF00) | v as u16;
    }
    fn set_hi(word: &mut u16, v: u8) {
        *word = (*word & 0x00FF) | ((v as u16) << 8);
    }
    fn set_lo_i(word: &mut i16, v: u8) {
        *word = (*word as u16 & 0xFF00 | v as u16) as i16;
    }
    fn set_hi_i(word: &mut i16, v: u8) {
        *word = (*word as u16 & 0x00FF | ((v as u16) << 8)) as i16;
    }
    match offset {
        4 => set_lo(&mut al.shape, value),
        5 => set_hi(&mut al.shape, value),
        6 => set_lo(&mut al.ptr, value),
        7 => set_hi(&mut al.ptr, value),
        8 => al.flags = value,
        9 => al.type_ = value,
        10 => al.count = value,
        11 => al.count1 = value,
        12 => set_lo_i(&mut al.worldx, value),
        13 => set_hi_i(&mut al.worldx, value),
        14 => set_lo_i(&mut al.worldy, value),
        15 => set_hi_i(&mut al.worldy, value),
        16 => set_lo_i(&mut al.worldz, value),
        17 => set_hi_i(&mut al.worldz, value),
        18 => al.rotx = value,
        19 => al.roty = value,
        20 => al.rotz = value,
        21 => al.vel = value,
        // al_stratptr (22..24) denied — see module docs.
        25 => set_lo(&mut al.immuneptr, value),
        26 => set_hi(&mut al.immuneptr, value),
        27 => set_lo(&mut al.collobjptr, value),
        28 => set_hi(&mut al.collobjptr, value),
        29 => al.sflags = value,
        30 => al.sflags2 = value,
        31 => al.sflags3 = value,
        32 => al.sflags4 = value,
        33 => al.skidy = value,
        34 => al.sbyte1 = value,
        35 => al.sbyte2 = value,
        36 => al.sbyte3 = value,
        37 => al.sbyte4 = value,
        38 => set_lo_i(&mut al.sword1, value),
        39 => set_hi_i(&mut al.sword1, value),
        40 => set_lo_i(&mut al.sword2, value),
        41 => set_hi_i(&mut al.sword2, value),
        42 => al.hp = value,
        43 => al.ap = value,
        _ => return false,
    }
    true
}

/// Read one byte of an `alx_start`-relative slot (C `alien_alx_byte_ptr` read).
fn alx_read(al: &Alien, offset: u16) -> Option<u8> {
    Some(match offset {
        0 => al.swpx1.to_le_bytes()[0],
        1 => al.swpx1.to_le_bytes()[1],
        2 => al.swpy1.to_le_bytes()[0],
        3 => al.swpy1.to_le_bytes()[1],
        4 => al.swpz1.to_le_bytes()[0],
        5 => al.swpz1.to_le_bytes()[1],
        // alx strategy pointers (6..17 on SNES) are 24-bit function pointers; denied.
        18 => al.stratstate,
        19 => al.fireobjptr.to_le_bytes()[0],
        20 => al.fireobjptr.to_le_bytes()[1],
        21 => al.depthoffset.to_le_bytes()[0],
        22 => al.depthoffset.to_le_bytes()[1],
        23 => al.relposx,
        24 => al.relposy,
        25 => al.relposz,
        26 => al.debrisshape.to_le_bytes()[0],
        27 => al.debrisshape.to_le_bytes()[1],
        28 => al.colframe,
        29 => al.animframe,
        30 => al.snd1,
        31 => al.snd2,
        32 => al.coltab.to_le_bytes()[0],
        33 => al.coltab.to_le_bytes()[1],
        34 => al.childx,
        35 => al.childy,
        36 => al.childz,
        37 => al.childrotx,
        38 => al.childroty,
        39 => al.childrotz,
        40 => al.childrotobj.to_le_bytes()[0],
        41 => al.childrotobj.to_le_bytes()[1],
        42 => al.tx,
        43 => al.ty,
        44 => al.memptr.to_le_bytes()[0],
        45 => al.memptr.to_le_bytes()[1],
        46 => al.stackptr.to_le_bytes()[0],
        47 => al.stackptr.to_le_bytes()[1],
        48 => al.stratmem.to_le_bytes()[0],
        49 => al.stratmem.to_le_bytes()[1],
        50 => al.pbyte1,
        51 => al.pbyte2,
        52 => al.pword1.to_le_bytes()[0],
        53 => al.pword1.to_le_bytes()[1],
        _ => return None,
    })
}

/// Write one byte of an `alx_start`-relative slot (C `alien_alx_byte_ptr` write).
fn alx_write(al: &mut Alien, offset: u16, value: u8) -> bool {
    fn set_lo(word: &mut u16, v: u8) {
        *word = (*word & 0xFF00) | v as u16;
    }
    fn set_hi(word: &mut u16, v: u8) {
        *word = (*word & 0x00FF) | ((v as u16) << 8);
    }
    fn set_lo_i(word: &mut i16, v: u8) {
        *word = (*word as u16 & 0xFF00 | v as u16) as i16;
    }
    fn set_hi_i(word: &mut i16, v: u8) {
        *word = (*word as u16 & 0x00FF | ((v as u16) << 8)) as i16;
    }
    match offset {
        0 => set_lo_i(&mut al.swpx1, value),
        1 => set_hi_i(&mut al.swpx1, value),
        2 => set_lo_i(&mut al.swpy1, value),
        3 => set_hi_i(&mut al.swpy1, value),
        4 => set_lo_i(&mut al.swpz1, value),
        5 => set_hi_i(&mut al.swpz1, value),
        // alx strategy pointers (6..17) denied — see module docs.
        18 => al.stratstate = value,
        19 => set_lo(&mut al.fireobjptr, value),
        20 => set_hi(&mut al.fireobjptr, value),
        21 => set_lo_i(&mut al.depthoffset, value),
        22 => set_hi_i(&mut al.depthoffset, value),
        23 => al.relposx = value,
        24 => al.relposy = value,
        25 => al.relposz = value,
        26 => set_lo(&mut al.debrisshape, value),
        27 => set_hi(&mut al.debrisshape, value),
        28 => al.colframe = value,
        29 => al.animframe = value,
        30 => al.snd1 = value,
        31 => al.snd2 = value,
        32 => set_lo(&mut al.coltab, value),
        33 => set_hi(&mut al.coltab, value),
        34 => al.childx = value,
        35 => al.childy = value,
        36 => al.childz = value,
        37 => al.childrotx = value,
        38 => al.childroty = value,
        39 => al.childrotz = value,
        40 => set_lo(&mut al.childrotobj, value),
        41 => set_hi(&mut al.childrotobj, value),
        42 => al.tx = value,
        43 => al.ty = value,
        44 => set_lo(&mut al.memptr, value),
        45 => set_hi(&mut al.memptr, value),
        46 => set_lo(&mut al.stackptr, value),
        47 => set_hi(&mut al.stackptr, value),
        48 => set_lo(&mut al.stratmem, value),
        49 => set_hi(&mut al.stratmem, value),
        50 => al.pbyte1 = value,
        51 => al.pbyte2 = value,
        52 => set_lo(&mut al.pword1, value),
        53 => set_hi(&mut al.pword1, value),
        _ => return false,
    }
    true
}

/// C `AlienCompat_Read8` (src/game/alien_compat.c).
pub fn read8(al: &Alien, offset: u16, is_alx: bool) -> Option<u8> {
    if is_alx {
        alx_read(al, offset)
    } else {
        base_read(al, offset)
    }
}

/// C `AlienCompat_Read16` (src/game/alien_compat.c): two 8-bit reads, LE.
pub fn read16(al: &Alien, offset: u16, is_alx: bool) -> Option<u16> {
    let lo = read8(al, offset, is_alx)?;
    let hi = read8(al, offset.wrapping_add(1), is_alx)?;
    Some(u16::from(lo) | (u16::from(hi) << 8))
}

/// C `AlienCompat_Write8` (src/game/alien_compat.c).
pub fn write8(al: &mut Alien, offset: u16, is_alx: bool, value: u8) -> bool {
    if is_alx {
        alx_write(al, offset, value)
    } else {
        base_write(al, offset, value)
    }
}

/// C `AlienCompat_Write16` (src/game/alien_compat.c): two 8-bit writes, LE.
/// Like C, the low byte is committed even if the high byte slot is invalid.
pub fn write16(al: &mut Alien, offset: u16, is_alx: bool, value: u16) -> bool {
    if !write8(al, offset, is_alx, (value & 0xFF) as u8) {
        return false;
    }
    write8(al, offset.wrapping_add(1), is_alx, (value >> 8) as u8)
}

/// C `AlienCompat_Add8` — 8-bit wrapping add in place.
pub fn add8(al: &mut Alien, offset: u16, is_alx: bool, delta: i8) -> bool {
    let cur = match read8(al, offset, is_alx) {
        Some(v) => v,
        None => return false,
    };
    write8(al, offset, is_alx, cur.wrapping_add(delta as u8))
}

/// C `AlienCompat_Add16` — 16-bit wrapping add in place.
pub fn add16(al: &mut Alien, offset: u16, is_alx: bool, delta: i16) -> bool {
    let cur = match read16(al, offset, is_alx) {
        Some(v) => v,
        None => return false,
    };
    write16(al, offset, is_alx, cur.wrapping_add(delta as u16))
}

// Path bytecode variable offsets use PALVAROFFSET encoding:
// low 7 bits = offset, bit7 set = alx offset (alien_compat.h inline helpers).

/// C `AlienCompat_PathRead8` (src/game/alien_compat.h).
pub fn path_read8(al: &Alien, encoded_offset: u8) -> Option<u8> {
    read8(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0)
}

/// C `AlienCompat_PathRead16` (src/game/alien_compat.h).
pub fn path_read16(al: &Alien, encoded_offset: u8) -> Option<u16> {
    read16(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0)
}

/// C `AlienCompat_PathWrite8` (src/game/alien_compat.h).
pub fn path_write8(al: &mut Alien, encoded_offset: u8, value: u8) -> bool {
    write8(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0, value)
}

/// C `AlienCompat_PathWrite16` (src/game/alien_compat.h).
pub fn path_write16(al: &mut Alien, encoded_offset: u8, value: u16) -> bool {
    write16(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0, value)
}

/// C `AlienCompat_PathAdd8` (src/game/alien_compat.h).
pub fn path_add8(al: &mut Alien, encoded_offset: u8, delta: i8) -> bool {
    add8(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0, delta)
}

/// C `AlienCompat_PathAdd16` (src/game/alien_compat.h).
pub fn path_add16(al: &mut Alien, encoded_offset: u8, delta: i16) -> bool {
    add16(al, u16::from(encoded_offset & 0x7F), encoded_offset & 0x80 != 0, delta)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::alien::Alien;

    #[test]
    fn word_fields_are_little_endian() {
        let mut al = Alien::default();
        assert!(write16(&mut al, 12, false, 0xBEEF));
        assert_eq!(al.worldx as u16, 0xBEEF);
        assert_eq!(read8(&al, 12, false), Some(0xEF));
        assert_eq!(read8(&al, 13, false), Some(0xBE));
    }

    #[test]
    fn function_pointer_offsets_denied() {
        let al = Alien::default();
        for off in [0u16, 1, 2, 3, 22, 23, 24] {
            assert_eq!(read8(&al, off, false), None, "base offset {off}");
        }
        for off in 6u16..=17 {
            assert_eq!(read8(&al, off, true), None, "alx offset {off}");
        }
    }

    #[test]
    fn path_encoding_selects_alx() {
        let mut al = Alien::default();
        // PAL_PBYTE1 = 0x80 | 50
        assert!(path_write8(&mut al, 0x80 | 50, 7));
        assert_eq!(al.pbyte1, 7);
        // base offset 42 = HP
        assert!(path_write8(&mut al, 42, 9));
        assert_eq!(al.hp, 9);
    }
}
